### PR TITLE
Enable nullable reference types.

### LINKED ISSUE #50: Nullable reference types
Consider using the new nullable reference types features for more runtime robustness. 
To get started:

- https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/
- https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references

### PR DESCRIPTION
**Pull request by NikolajDL**
fixes #50

Enables nullable reference types for the entire MonkeyBot project in the .csproj.
Re-aligns LangVersion-property.

- This change produces a lot of warnings (200+) that will need to be looked at.
- I have already started fixing some "low-hanging fruits" around the codebase, but there's potentially a lot of small changes to a lot of files, so it'll be quite a hassle to review.

### CODE REVIEW

File: MonkeyBot/Database/MonkeyDBContext.cs [hunk 1]
```diff
@@ -21,13 +21,15 @@ public class MonkeyDBContext : DbContext
         public DbSet<GameSubscription> GameSubscriptions { get; set; }
         public DbSet<RoleButtonLink> RoleButtonLinks { get; set; }
 
+#pragma warning disable CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
```
Comment (Excolo) [inline MonkeyBot/Database/MonkeyDBContext.cs]: I didn't want to mess with making the EF context DbSet's nullable - nor handle nullability anywhere it's used, since I know it'll be assigned by EF. Instead I just disabled the warning for this lone instance.

File: MonkeyBot/Common/CloudinaryCredentials.cs [hunk 2]
```diff
@@ -2,8 +2,8 @@
 {
     public class CloudinaryCredentials
     {
-        public string Cloud { get; set; }
-        public string ApiKey { get; set; }
-        public string ApiSecret { get; set; }
+        public string? Cloud { get; set; }
+        public string? ApiKey { get; set; }
+        public string? ApiSecret { get; set; }
```
Comment (MarkusKgit) [inline MonkeyBot/Common/CloudinaryCredentials.cs]: Not sure if these should be nullable. It would probably be better to have a constructor that sets these properties and then they could also be readonly.

File: MonkeyBot/CommandManager.cs [hunk 3]
```diff
@@ -1,4 +1,5 @@
-﻿using Discord;
+﻿
```
Comment (MarkusKgit) [inline MonkeyBot/CommandManager.cs]: nit: unnecessary blank line

File: MonkeyBot/Common/MonkeyHelpers.cs [hunk 4]
```diff
@@ -19,8 +19,8 @@ public static async Task WriteTextAsync(string filePath, string text, bool appen
         {
             if (!File.Exists(filePath))
             {
-                string strippedPath = Path.GetDirectoryName(filePath);
-                if (!strippedPath.IsEmpty() && !Directory.Exists(strippedPath))
+                string? strippedPath = Path.GetDirectoryName(filePath);
+                if (strippedPath != null && !strippedPath.IsEmpty() && !Directory.Exists(strippedPath))
```
Comment (MarkusKgit) [inline MonkeyBot/Common/MonkeyHelpers.cs]: IsEmpty() is a shorthand for string.IsNullOrEmpty(), I wonder if there's a way to make the analyzer pick that up?

File: MonkeyBot/Common/MonkeyHelpers.cs [hunk 5]
```diff
@@ -69,17 +69,23 @@ public static async Task<string> ReadTextAsync(string filePath)
         /// <param name="guildID">Id of the Discord guild</param>
         /// <param name="channelID">Id of the Discord channel</param>
         /// <param name="text">Text to post</param>
-        public static async Task<IUserMessage> SendChannelMessageAsync(IDiscordClient client, ulong guildID, ulong channelID, string text, bool isTTS = false, Embed embed = null, RequestOptions options = null)
+        public static async Task<IUserMessage> SendChannelMessageAsync(IDiscordClient client, ulong guildID, ulong channelID, string text, bool isTTS = false, Embed? embed = null, RequestOptions? options = null)
         {
-            IGuild guild = await (client?.GetGuildAsync(guildID)).ConfigureAwait(false);
-            ITextChannel channel = await (guild?.GetTextChannelAsync(channelID)).ConfigureAwait(false);
-            return await (channel?.SendMessageAsync(text, isTTS, embed, options)).ConfigureAwait(false);
+            IGuild guild = await client.GetGuildAsync(guildID).ConfigureAwait(false);
```
Comment (MarkusKgit) [inline MonkeyBot/Common/MonkeyHelpers.cs]: GetGuildAsync can return null

File: MonkeyBot/Common/MonkeyHelpers.cs [hunk 6]
```diff
@@ -69,17 +69,23 @@ public static async Task<string> ReadTextAsync(string filePath)
...
+            ITextChannel channel = await (guild.GetTextChannelAsync(channelID)).ConfigureAwait(false);
```
Comment (MarkusKgit) [inline MonkeyBot/Common/MonkeyHelpers.cs]: GetTextChannelAsync can return null

File: MonkeyBot/Common/MonkeyHelpers.cs [hunk 7]
```diff
@@ -69,17 +69,23 @@ public static async Task<string> ReadTextAsync(string filePath)
         /// <param name="guildID">Id of the Discord guild</param>
         /// <param name="channelID">Id of the Discord channel</param>
         /// <param name="text">Text to post</param>
-        public static async Task<IUserMessage> SendChannelMessageAsync(IDiscordClient client, ulong guildID, ulong channelID, string text, bool isTTS = false, Embed embed = null, RequestOptions options = null)
+        public static async Task<IUserMessage> SendChannelMessageAsync(IDiscordClient client, ulong guildID, ulong channelID, string text, bool isTTS = false, Embed? embed = null, RequestOptions? options = null)
```
Comment (MarkusKgit) [inline MonkeyBot/Common/MonkeyHelpers.cs]: client could also be null

File: MonkeyBot/Common/MonkeyModuleBase.cs [hunk 8]
```diff
@@ -9,29 +9,29 @@ namespace MonkeyBot.Common
...
-            IGuildUser user = null;
+            IGuildUser? user = null;
             if (userName.StartsWith("<@", StringComparison.InvariantCulture) && ulong.TryParse(userName.Replace("<@", "", StringComparison.InvariantCulture).Replace(">", "", StringComparison.InvariantCulture), out ulong id))
             {
                 user = await Context.Guild.GetUserAsync(id).ConfigureAwait(false);
             }
             else
             {
-                IEnumerable<IGuildUser> users = (await (Context.Guild?.GetUsersAsync()).ConfigureAwait(false))?.Where(x => x.Username.Contains(userName, StringComparison.OrdinalIgnoreCase));
+                IEnumerable<IGuildUser> users = (await Context.Guild.GetUsersAsync().ConfigureAwait(false)).Where(x => x.Username.Contains(userName, StringComparison.OrdinalIgnoreCase));
```
Comment (MarkusKgit) [inline MonkeyBot/Common/MonkeyModuleBase.cs]: what's the reason behind these changes?

File: MonkeyBot/Common/XYChart.cs [hunk 9]
```diff
@@ -45,6 +45,8 @@ public XYChart(int width, int height, int borderMargin)
             chartWidth = width;
             chartHeight = height;
             axisBorder = borderMargin;
+            AxisX = axisX;
```
Comment (MarkusKgit) [inline MonkeyBot/Common/XYChart.cs]: does this not complain about AxisX/Y being possibly null?

File: MonkeyBot/Database/MonkeyDBContext.cs (hunk 1 above)
Comment (MarkusKgit) [reply MonkeyBot/Database/MonkeyDBContext.cs]: good catch

File: MonkeyBot/Documentation/DocumentationBuilder.cs [hunk 10]
```diff
@@ -128,7 +128,7 @@ private static string TranslatePrecondition(PreconditionAttribute precondition,
             }
             else
             {
-                return precondition.ToString();
+                return precondition.ErrorMessage;
```
Comment (MarkusKgit) [inline MonkeyBot/Documentation/DocumentationBuilder.cs]: why this change?

File: MonkeyBot/Initializer.cs [hunk 11]
```diff
@@ -27,7 +27,7 @@ public static async Task<IServiceProvider> InitializeAsync(ApplicationArguments
             });
 
             DiscordSocketClient client = services.GetService<DiscordSocketClient>();
-            string token = (await DiscordClientConfiguration.LoadAsync().ConfigureAwait(false)).Token;
+            string? token = (await DiscordClientConfiguration.LoadAsync().ConfigureAwait(false)).Token;
             await client.LoginAsync(TokenType.Bot, token).ConfigureAwait(false);
```
Comment (MarkusKgit) [inline MonkeyBot/Initializer.cs]: maybe we should assert the token first before trying to Log In (neither null nor empty string)

**Comment by MarkusKgit** [comment]
also could you please rebase onto master, there were some changes to the JSON usage recently

File: MonkeyBot/Common/CloudinaryCredentials.cs (hunk 2 above)
Comment (Excolo) [reply MonkeyBot/Common/CloudinaryCredentials.cs]: I was thinking that too, but didn't want to mess too much with the code where these properties are assigned, without knowing exactly the original thought behind it.

File: MonkeyBot/Common/MonkeyHelpers.cs (hunk 4 above)
Comment (Excolo) [reply MonkeyBot/Common/MonkeyHelpers.cs]: It might be fixed by changing the extension to extend a nullable string `IsEmpty(this string? s)`. I didn't look into it too deep, so I didn't catch that it also checked for null. I'm a strong proponent of verbose naming, so I just assumed it only checked for emptyness and not null. :-)

File: MonkeyBot/Common/MonkeyHelpers.cs (hunk 7 above)
Comment (Excolo) [reply MonkeyBot/Common/MonkeyHelpers.cs]: Yeah, I guess everything from the Discord API is nullable.

File: MonkeyBot/Common/XYChart.cs (hunk 9 above)
Comment (Excolo) [reply MonkeyBot/Common/XYChart.cs]: Nope, because you cannot pass null into the constructor without getting a similar warning. :-)

File: MonkeyBot/Documentation/DocumentationBuilder.cs (hunk 10 above)
Comment (Excolo) [reply MonkeyBot/Documentation/DocumentationBuilder.cs]: I wanted the methods return type to be non-nullable (otherwise I would introduce a cascade of new warnings), but `ToString()` for some reason returns nullable string, so I wanted to use something similar and noticed that the error message wasn't nullable. I was definitely a bit uncertain of the correctness of this fix - especially not having a setup to run the bot myself. :-)

File: MonkeyBot/Initializer.cs (hunk 11 above)
Comment (Excolo) [reply MonkeyBot/Initializer.cs]: Agreed

File: MonkeyBot/Common/MonkeyHelpers.cs (hunk 4 above)
Comment (MarkusKgit) [reply MonkeyBot/Common/MonkeyHelpers.cs]: Agreed on the naming, I was being inconsistent there :) Will change it later

File: MonkeyBot/Initializer.cs (hunk 11 above)
Comment (MarkusKgit) [reply MonkeyBot/Initializer.cs]: ErrorMessage is only set when the Precondition is actually checked before running a command if I interpret it correctly. It defaults to null. 
But just leave as is, I just looked at the whole method and it just asks for a refactoring to a switch expression anyway :)

File: MonkeyBot/Documentation/DocumentationBuilder.cs (hunk 10 above)
Comment (MarkusKgit) [reply MonkeyBot/Documentation/DocumentationBuilder.cs]: ErrorMessage is only set when the Precondition is actually checked before running a command if I interpret it correctly. It defaults to null.
But just leave as is, I just looked at the whole method and it just asks for a refactoring to a switch expression anyway :)

File: MonkeyBot/Initializer.cs (hunk 11 above)
Comment (MarkusKgit) [reply MonkeyBot/Initializer.cs]: I think it throws when it can't log in but I will have a look after merging. It's a bit off scope here

File: MonkeyBot/Initializer.cs [hunk 12]
```diff
@@ -28,6 +28,16 @@ public static async Task<IServiceProvider> InitializeAsync(ApplicationArguments
 
             DiscordSocketClient client = services.GetService<DiscordSocketClient>();
             string? token = (await DiscordClientConfiguration.LoadAsync().ConfigureAwait(false)).Token;
+            try
```
Comment (MarkusKgit) [inline MonkeyBot/Initializer.cs]: Since I was digging in the Discord API anyway I decided to just update this in the meantime ;)
